File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
# TC- O(logn)
# SC- O(1)

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # we can do binary search for efficiency. But to figure out which half we need to search, we need to first identify which half is sorted. Once you identify teh sorted half, you compare with target value in that half.
        left, right = 0, len(nums)-1

        while left<=right:
            mid = left+(right-left)//2

            if nums[mid]==target:
                return mid
            
            # left half is sorted
            if nums[left]<=nums[mid]:
                if nums[left]<=target<nums[mid]:
                    right = mid-1
                else:
                    left = mid+1
            
            # right half (including mid) is not sorted
            else:
                if nums[mid]<target<=nums[right]:
                    left = mid+1
                else:
                    right = mid-1
        return -1

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cpp
// TC- O(logn)
// SC- O(1)

class Solution {
public:
    int search(vector<int>& nums, int target) {
        // 0 1 2 4 5 6 7 - check left if not rotated
        // 4 5 6 7 0 1 2 - check right
        // 7 0 1 2 4 5 6 - check left
        // 3 4 5 6 7 1 2 - 

        // if only one element
        if(nums.size()==1 and nums[0]==target)
            return 0;
        
        int l=0;
        int r=nums.size()-1;
        while(l<=r){
            int mid = l+ (r-l)/2;
            cout<<l<<" "<<r<<" "<<mid<<endl;
            if(nums[mid]==target)
                return mid;
            else if(nums[l]<=nums[mid]){
                // left subarray is sorted
                if(target<nums[mid] and target>=nums[l])
                    r = mid-1;
                else
                    l = mid+1;
            }
            else
                // right subarray is sorted
                if(target>nums[mid] and target<=nums[r])
                    l = mid+1;
                else
                    r = mid-1;
        }
        return -1;
    }
};